packages/cosmos/src/client.rs

```diff
@@ -1270,91 +1300,131 @@ impl TxBuilder {
-        }
-        .tx_response
-        .context("Missing inner tx_response")?;
-
-        if !self.skip_code_check && res.code != 0 {
-            let e = anyhow::anyhow!(
-                "Initial transaction broadcast failed with code {}. Raw log: {}",
-                res.code,
-                res.raw_log
-            );
-            let is_sequence = get_expected_sequence(&res.raw_log);
-            return Err(match is_sequence {
-                None => ExpectedSequenceError::RealError(e),
-                Some(number) => ExpectedSequenceError::NewNumber(number, e),
-            });
-        };
+            .tx_response
+            .context("Missing inner tx_response")?;
+
+            if !self.skip_code_check && res.code != 0 {
+                let e = anyhow::anyhow!(
+                    "Initial transaction broadcast failed with code {}. Raw log: {}",
+                    res.code,
+                    res.raw_log
+                );
+                if res.code == 13 {
+                    return Err(AttemptError::InsufficientGas(e));
+                }
+                let is_sequence = get_expected_sequence(&res.raw_log);
+                return Err(AttemptError::Inner(match is_sequence {
+                    None => ExpectedSequenceError::RealError(e),
+                    Some(number) => ExpectedSequenceError::NewNumber(number, e),
+                }));
+            };
 
-        log::debug!("Initial BroadcastTxResponse: {res:?}");
-
-        let res = cosmos.wait_for_transaction(res.txhash).await?;
-        if !self.skip_code_check && res.code != 0 {
-            // We don't do the account sequence mismatch hack work here, once a
-            // transaction actually lands on the chain we don't want to ever
-            // automatically retry.
-            return Err(ExpectedSequenceError::RealError(anyhow::anyhow!(
-                "Transaction failed with code {}. Raw log: {}",
-                res.code,
-                res.raw_log
-            )));
+            log::debug!("Initial BroadcastTxResponse: {res:?}");
+
+            let res = cosmos.wait_for_transaction(res.txhash).await?;
+            if !self.skip_code_check && res.code != 0 {
+                // We don't do the account sequence mismatch hack work here, once a
+                // transaction actually lands on the chain we don't want to ever
+                // automatically retry.
+                return Err(AttemptError::Inner(ExpectedSequenceError::RealError(
+                    anyhow::anyhow!(
+                        "Transaction failed with code {}. Raw log: {}",
+                        res.code,
+                        res.raw_log
+                    ),
+                )));
+            };
+
+            log::debug!("TxResponse: {res:?}");
+
+            Ok(res)
         };
 
-        log::debug!("TxResponse: {res:?}");
+        let attempts = cosmos.get_first_builder().config.gas_price_retry_attempts;
+        for attempt_number in 0..attempts {
```

Comment:
I'm not 100% sure, but if this has an inclusive max (i.e. `0..=attempts`), wouldn't it still work due to the `>=` in `retry_with_price()` and then the last attempt outside of the for loop can be avoided, if we get there it's an error?

If my question/suggestion here isn't clear, lmk

Reply:
You're right about all that I believe. My reason for exclusive max was so that I could have the final attempt outside the loop and return an error message.